### PR TITLE
Significantly reduce thrash allocations in zipkin encoding

### PR DESCRIPTION
**Pull request by johnbley**
* Use a short-circuit for turning 64-bit IDs into hex
* Remove some spurious extra allocations/doubled work
* Rework buffer strategy between serialization and writing to the wire to not double-copy traces.

### CODE REVIEW

File: dd-trace-api/src/main/java/datadog/trace/api/Ids.java [hunk 1]
```diff
@@ -5,9 +5,31 @@
...
+  };
 
   public static char[] idToHexChars(final String id) {
+    if (id.length() <= 20) {
+      // *likely* (but not definitely) a 64-bit id: try to parse as such
+      try {
+        long val = Long.parseLong(id);
+        char[] answer = new char[16];
+        for (int i = 0; i < 16; i++) {
+          long mask = (0x0FL << (i * 4));
```
Comment (rmfitzpatrick) [inline dd-trace-api/src/main/java/datadog/trace/api/Ids.java:23]: these can be precomputed or will the jvm likely do that?

File: dd-trace-ot/src/main/java/datadog/trace/common/writer/ZipkinV2Api.java [hunk 2]
```diff
@@ -67,22 +66,24 @@ public Response sendTraces(final List<List<DDSpan>> traces) {
...
-  public byte[] serializeTrace(final List<DDSpan> trace) throws IOException {
-    final ByteArrayOutputStream stream = new ByteArrayOutputStream(trace.size() * 128);
+  public SerializedBuffer serializeTrace(final List<DDSpan> trace) throws IOException {
+    final StreamingSerializedBuffer stream = new StreamingSerializedBuffer(trace.size() * 128);
     final JsonGenerator jsonGenerator = JSON_FACTORY.createGenerator(stream, JsonEncoding.UTF8);
 
     jsonGenerator.writeStartArray();
-    for (DDSpan span : trace) {
-      writeSpan(span, jsonGenerator);
+    for (int i = 0; i < trace.size(); i++) {
```
Comment (rmfitzpatrick) [inline dd-trace-ot/src/main/java/datadog/trace/common/writer/ZipkinV2Api.java:74]: Is this iterator adding overhead?

File: dd-trace-ot/src/main/java/datadog/trace/common/writer/ZipkinV2Api.java [hunk 3]
```diff
@@ -91,14 +92,14 @@ public Response sendSerializedTraces(
         int traceCount = 0;
 
         out.write('[');
```
Comment (rmfitzpatrick) [inline dd-trace-ot/src/main/java/datadog/trace/common/writer/ZipkinV2Api.java:94]: This has always made me feel a little dirty.  Would using the streaming api be better in your opinion?

File: dd-trace-api/src/main/java/datadog/trace/api/Ids.java (hunk 1 above)
Comment (johnbley) [reply dd-trace-api/src/main/java/datadog/trace/api/Ids.java:23]: Not worth it; a move is so much more expensive than a one-op shift.  Also, the memory allocation saved here outweighs either by several orders of magnitude.

File: dd-trace-ot/src/main/java/datadog/trace/common/writer/ZipkinV2Api.java (hunk 2 above)
Comment (johnbley) [reply dd-trace-ot/src/main/java/datadog/trace/common/writer/ZipkinV2Api.java:74]: It's a tiny piece of the overall puzzle and almost certainly elided by escape analysis, but every bit matters for this particular allocation hot path.

File: dd-trace-ot/src/main/java/datadog/trace/common/writer/ZipkinV2Api.java (hunk 3 above)
Comment (johnbley) [reply dd-trace-ot/src/main/java/datadog/trace/common/writer/ZipkinV2Api.java:94]: I wouldn't lose any sleep over it; this code is EOL eventually anyway.

**Pull request merged**